File: lib/features/predictions/domain/prediction_history.dart
import 'cycle_prediction.dart';

class PredictionOutcome {
  const PredictionOutcome({
    required this.prediction,
    required this.actualStart,
  });

  final CyclePrediction prediction;
  final DateTime actualStart;

  int get signedErrorDays =>
      _date(actualStart).difference(_date(prediction.mostLikelyDate)).inDays;

  int get absoluteErrorDays => signedErrorDays.abs();

  bool get withinWindow {
    final actual = _date(actualStart);
    return !actual.isBefore(_date(prediction.windowStart)) &&
        !actual.isAfter(_date(prediction.windowEnd));
  }

  static DateTime _date(DateTime value) =>
      DateTime(value.year, value.month, value.day);
}

class PredictionEvaluation {
  const PredictionEvaluation({
    required this.sampleCount,
    required this.meanAbsoluteErrorDays,
    required this.medianAbsoluteErrorDays,
    required this.windowCoverage,
    required this.earlyBiasDays,
    required this.lateBiasDays,
    required this.overConfidenceRate,
    required this.underConfidenceRate,
  });

  final int sampleCount;
  final double meanAbsoluteErrorDays;
  final double medianAbsoluteErrorDays;
  final double windowCoverage;
  final double earlyBiasDays;
  final double lateBiasDays;
  final double overConfidenceRate;
  final double underConfidenceRate;
}

class PredictionEvaluator {
  const PredictionEvaluator();

  PredictionEvaluation evaluate(List<PredictionOutcome> outcomes) {
    if (outcomes.isEmpty) {
      return const PredictionEvaluation(
        sampleCount: 0,
        meanAbsoluteErrorDays: 0,
        medianAbsoluteErrorDays: 0,
        windowCoverage: 0,
        earlyBiasDays: 0,
        lateBiasDays: 0,
        overConfidenceRate: 0,
        underConfidenceRate: 0,
      );
    }

    final absolute = outcomes
        .map((value) => value.absoluteErrorDays.toDouble())
        .toList(growable: false);
    final signed = outcomes
        .map((value) => value.signedErrorDays.toDouble())
        .toList(growable: false);
    final early = signed
        .where((value) => value < 0)
        .map((value) => value.abs())
        .toList();
    final late = signed.where((value) => value > 0).toList();
    final highConfidence = outcomes
        .where(
          (value) => value.prediction.confidence == PredictionConfidence.high,
        )
        .toList(growable: false);
    final lowConfidence = outcomes
        .where(
          (value) => value.prediction.confidence == PredictionConfidence.low,
        )
        .toList(growable: false);

    return PredictionEvaluation(
      sampleCount: outcomes.length,
      meanAbsoluteErrorDays: absolute.reduce((a, b) => a + b) / absolute.length,
      medianAbsoluteErrorDays: _median(absolute),
      windowCoverage:
          outcomes.where((value) => value.withinWindow).length /
          outcomes.length,
      earlyBiasDays: early.isEmpty
          ? 0
          : early.reduce((a, b) => a + b) / early.length,
      lateBiasDays: late.isEmpty
          ? 0
          : late.reduce((a, b) => a + b) / late.length,
      overConfidenceRate: highConfidence.isEmpty
          ? 0
          : highConfidence.where((value) => !value.withinWindow).length /
                highConfidence.length,
      underConfidenceRate: lowConfidence.isEmpty
          ? 0
          : lowConfidence.where((value) => value.withinWindow).length /
                lowConfidence.length,
    );
  }

  double _median(List<double> values) {
    final sorted = [...values]..sort();
    final middle = sorted.length ~/ 2;
    if (sorted.length.isOdd) return sorted[middle];
    return (sorted[middle - 1] + sorted[middle]) / 2;
  }
}

File: web/src/domain/prediction/prediction-history.ts
import type { PredictionResult } from '../../features/predictions/prediction-engine';

const DAY_MS = 86_400_000;

export type PredictionOutcome = {
  prediction: PredictionResult;
  actualStart: string;
};

export type PredictionEvaluation = {
  sampleCount: number;
  meanAbsoluteErrorDays: number;
  medianAbsoluteErrorDays: number;
  windowCoverage: number;
  earlyBiasDays: number;
  lateBiasDays: number;
  overConfidenceRate: number;
  underConfidenceRate: number;
};

function dateOnlyMs(value: string): number {
  const date = value.slice(0, 10);
  const match = /^(\d{4})-(\d{2})-(\d{2})$/.exec(date);
  if (!match) throw new Error(`prediction history requires a calendar date: ${value}`);
  const result = Date.UTC(Number(match[1]), Number(match[2]) - 1, Number(match[3]));
  if (new Date(result).toISOString().slice(0, 10) !== date) {
    throw new Error(`prediction history received an invalid calendar date: ${value}`);
  }
  return result;
}

function signedErrorDays(outcome: PredictionOutcome): number {
  return Math.round(
    (dateOnlyMs(outcome.actualStart) - dateOnlyMs(outcome.prediction.mostLikelyDate)) / DAY_MS,
  );
}

function withinWindow(outcome: PredictionOutcome): boolean {
  const actual = dateOnlyMs(outcome.actualStart);
  return actual >= dateOnlyMs(outcome.prediction.windowStart)
    && actual <= dateOnlyMs(outcome.prediction.windowEnd);
}

function mean(values: number[]): number {
  if (values.length === 0) return 0;
  return values.reduce((sum, value) => sum + value, 0) / values.length;
}

function median(values: number[]): number {
  if (values.length === 0) return 0;
  const ordered = [...values].sort((left, right) => left - right);
  const middle = Math.floor(ordered.length / 2);
  return ordered.length % 2 === 1
    ? ordered[middle]
    : (ordered[middle - 1] + ordered[middle]) / 2;
}

export function evaluatePredictionHistory(
  outcomes: PredictionOutcome[],
): PredictionEvaluation {
  if (outcomes.length === 0) {
    return {
      sampleCount: 0,
      meanAbsoluteErrorDays: 0,
      medianAbsoluteErrorDays: 0,
      windowCoverage: 0,
      earlyBiasDays: 0,
      lateBiasDays: 0,
      overConfidenceRate: 0,
      underConfidenceRate: 0,
    };
  }

  const signed = outcomes.map(signedErrorDays);
  const absolute = signed.map((value) => Math.abs(value));
  const early = signed.filter((value) => value < 0).map((value) => Math.abs(value));
  const late = signed.filter((value) => value > 0);
  const highConfidence = outcomes.filter((outcome) => outcome.prediction.confidence === 'high');
  const lowConfidence = outcomes.filter((outcome) => outcome.prediction.confidence === 'low');

  return {
    sampleCount: outcomes.length,
    meanAbsoluteErrorDays: mean(absolute),
    medianAbsoluteErrorDays: median(absolute),
    windowCoverage: outcomes.filter(withinWindow).length / outcomes.length,
    earlyBiasDays: mean(early),
    lateBiasDays: mean(late),
    overConfidenceRate: highConfidence.length === 0
      ? 0
      : highConfidence.filter((outcome) => !withinWindow(outcome)).length / highConfidence.length,
    underConfidenceRate: lowConfidence.length === 0
      ? 0
      : lowConfidence.filter(withinWindow).length / lowConfidence.length,
  };
}
